pilot/pkg/proxy/envoy/v2/debounced_channel.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	"sync"
+	"time"
+)
+
+// DebouncedEvent is used as entry type, which is enqueued in DebouncedChannel
+type DebouncedEvent = bool
+
+// DebouncedChannel is a channel which abstracts the debounce algorithm
+type DebouncedChannel struct {
+	// C channel, which can be used in selects. It delivers debounced events.
+	// The function emitted must be called. Otherwise processing is stopped.
+	C <-chan func(func(DebouncedEvent))
+	c chan func(func(DebouncedEvent))
```

Comment:
comments on these two channel

Reply:
the channel type is so complicated

Reply:
> comments on these two channel

See: https://github.com/istio/istio/pull/15810/files/513e1ea0447323a7ae0190fc279acb3068c0e5b9#r308665564
tl;dr Same instance, one reference is public but read-only, the other private rw. 



> the channel type is so complicated

The "callback inside a callback" allows us to track the execution time of the job without any help of the calling code.
I'm not aware of a better solution. Any suggestions from your side @hzxuzhonghu?

Reply:
I donot think too much